### PR TITLE
Update to sqlparser 0.14

### PR DESCRIPTION
**Pull request by alamb**
~Draft until sqlparser 0.14 is actually released~

# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/pull/1804

 # Rationale for this change

A new version of sqlparser is about to be released, and this PR serves as potential QA for this effort

# What changes are included in this PR?
1. Update sqlparser version
2. update datafusion sqlparser for new changes

# Are there any user-facing changes?
New dependency and some new SQL syntax support

### CODE REVIEW

File: datafusion/src/sql/planner.rs [hunk 1]
```diff
@@ -1241,11 +1246,20 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
     ) -> Result<Expr> {
-        match sql {
-            FunctionArg::Named { name: _, arg } => {
-                self.sql_expr_to_logical_expr(arg, schema)
+        let arg: &FunctionArgExpr = match sql {
+            FunctionArg::Named { name: _, arg } => arg,
+            FunctionArg::Unnamed(arg) => arg,
+        };
+
+        match arg {
```
Comment (alamb) [inline datafusion/src/sql/planner.rs:1254]: Needed due to https://github.com/sqlparser-rs/sqlparser-rs/pull/378 from @panarch  (so we don't allow crazy stuff like `SELECT * + * from foo` 👍

File: datafusion/src/sql/planner.rs [hunk 2]
```diff
@@ -85,30 +86,40 @@ pub struct SqlToRel<'a, S: ContextProvider> {
...
 
-#[allow(clippy::branches_sharing_code)]
-fn plan_indexed(expr: Expr, mut keys: Vec<Value>) -> Expr {
-    if keys.len() == 1 {
-        let key = keys.pop().unwrap();
-        Expr::GetIndexedField {
-            expr: Box::new(expr),
-            key: plan_key(key),
-        }
+fn plan_indexed(expr: Expr, mut keys: Vec<SQLExpr>) -> Result<Expr> {
```
Comment (alamb) [inline datafusion/src/sql/planner.rs:106]: I had to change this code anyways to handle arbitrary expressions due to https://github.com/sqlparser-rs/sqlparser-rs/pull/382 so I took the time to remove the clippy warning too

**Comment by alamb** [comment]
I think it might be nice to update to latest sqlparser before the 7.0.0 release

**Review by xudong963 — APPROVED** [review]
LGTM, I also think it would be better to join 7.0.0

**Comment by alamb** [comment]
Thanks @xudong963

**Pull request merged**